### PR TITLE
NVDRV & IHOSBinder Refactor + IOCTL2/IOCTL3 + Surface Deswizzling Fix

### PR DESCRIPTION
**Pull request by PixelyIon**
* Fix Surface Deswizzling OOB writes
* Move NVDRV + IHOSBinder Internals to Discrete Components
* Constexpr Maps for NvDevice IOCTLs
* NVDRV IOCTL Refactor
* * Buffer -> Span 
* * All buffers as arguments 
* * Return -> NvStatus 
* * Print Device Names + Function Names
* Implement IOCTL2 & IOCTL3

### CODE REVIEW

File: app/src/main/cpp/skyline/common.h [hunk 1]
```diff
@@ -194,6 +197,24 @@ namespace skyline {
             }
             return result;
         }
+
+        constexpr std::size_t Hash(std::string_view view) {
+            return frz::elsa<frz::string>{}(frz::string(view.data(), view.size()), 0);
+        }
+
+        template<typename Out, typename In>
+        constexpr Out& As(const std::span<In> &span) {
```
Comment (bylaws) [inline app/src/main/cpp/skyline/common.h:206]: Span reference

File: app/src/main/cpp/skyline/common.h [hunk 2]
```diff
@@ -194,6 +197,24 @@ namespace skyline {
...
+            if (span.size_bytes() < sizeof(Out))
+                throw exception("Span size less than Out type size");
+            return *reinterpret_cast<Out*>(span.data());
+        }
+
+        template<typename Out, typename In>
+        constexpr std::span<Out> AsSpan(const std::span<In> &span) {
+            if (span.size_bytes() < sizeof(Out))
```
Comment (bylaws) [inline app/src/main/cpp/skyline/common.h:214]: I think this is fine not to have, empty spans are a thing

File: app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        };
+        out.Push(output);
+
+        state.logger->Debug("QueueBuffer: Timestamp: {}, Auto Timestamp: {}, Crop: [T: {}, B: {}, L: {}, R: {}], Scaling Mode: {}, Transform: {}, Sticky Transform: {}, Swap Interval: {}, Slot: {}", data.timestamp, data.autoTimestamp, data.crop.top, data.crop.bottom, data.crop.left, data.crop.right, data.scalingMode, data.transform, data.stickyTransform, data.swapInterval, data.slot);
+    }
+
+    void GraphicBufferProducer::CancelBuffer(Parcel &in) {
+        u32 slot{in.Pop<u32>()};
+        //auto fences{in.Pop<std::array<nvdrv::Fence, 4>>()};
+
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:96]: What's up with the line a above this

File: app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                SetPreallocatedBuffer(in);
+                break;
+            default:
+                throw exception("An unimplemented transaction was called: {}", static_cast<u32>(code));
+        }
+    }
+
+    void GraphicBufferProducer::SetDisplay(const std::string &name) {
+        try {
+            const auto type = DisplayTypeMap.at(name);
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:195]: Const ref?

File: app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      public:
+        BufferStatus status{BufferStatus::Free};
+        std::shared_ptr<gpu::PresentationTexture> texture;
+        GbpBuffer gbpBuffer;
+
+        Buffer(const GbpBuffer &gbpBuffer, const std::shared_ptr<gpu::PresentationTexture> &texture);
+    };
+
+    /**
+     * @brief nvnflinger:dispdrv or nns::hosbinder::IHOSBinderDriver is responsible for writing buffers to the display
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h:54]: Change up this comment

File: app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            QueueBuffer = 7, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#353
+            CancelBuffer = 8, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#364
+            Query = 9, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#372
+            Connect = 10, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#381
+            Disconnect = 11, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#396
+            SetSidebandStream = 12, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#403
+            AllocateBuffers = 13, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#413
+            SetPreallocatedBuffer = 14, //!< No source on this but it's used to set a existing buffer according to libtransistor and libNX
+        };
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h:114]: s/libNX/libnx

File: app/src/main/cpp/skyline/services/hosbinder/IHOSBinderDriver.h [hunk 7]
```diff
@@ -5,87 +5,18 @@
...
-        Connect = 10, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#381
-        Disconnect = 11, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#396
-        SetSidebandStream = 12, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#403
-        AllocateBuffers = 13, //!< https://android.googlesource.com/platform/frameworks/native/+/8dc5539/libs/gui/IGraphicBufferProducer.cpp#413
-        SetPreallocatedBuffer = 14, //!< No source on this but it's used to set a existing buffer according to libtransistor and libNX
-    };
+    class GraphicBufferProducer;
 
     /**
      * @brief nvnflinger:dispdrv or nns::hosbinder::IHOSBinderDriver is responsible for writing buffers to the display
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/hosbinder/IHOSBinderDriver.h:13]: Redo comment

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvdevice.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+#include "nvdevice.h"
+
+namespace skyline::service::nvdrv::device {
+    const std::string &NvDevice::GetName() {
+        if (name.empty()) {
+            auto mangledName = typeid(*this).name();
+
+            int status{};
+            size_t length{};
+            std::unique_ptr<char, decltype(&std::free)> demangled{abi::__cxa_demangle(mangledName, nullptr, &length, &status), std::free};
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/nvdrv/devices/nvdevice.cpp:13]: Can we make a demangling util func?

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp [hunk 9]
```diff
@@ -43,107 +36,120 @@ namespace skyline::service::nvdrv::device {
...
+            state.logger->Warn("Invalid NvMap handle: 0x{:X}", data.nvmapHandle);
+            return NvStatus::BadParameter;
         }
-
-        state.process->WriteMemory(region, buffer.output.at(0).address);
     }
 
-    void NvHostAsGpu::GetVaRegions(IoctlData &buffer) {
+    NvStatus NvHostAsGpu::GetVaRegions(IoctlType type, std::span<u8> buffer, std::span<u8> inlineBuffer) {
+        /*
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp:91]: Whys this commented out?

File: app/src/main/cpp/skyline/services/nvdrv/driver.h [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    /**
+     * @brief nvnflinger:dispdrv or nns::hosbinder::IHOSBinderDriver is responsible for writing buffers to the display
+     */
+    class Driver {
+      private:
+        const DeviceState &state;
+        std::vector<std::shared_ptr<device::NvDevice>> devices; //!< A map from an FD to a shared pointer to it's NvDevice object
+        u32 fdIndex{}; //!< The index of a file descriptor
```
Comment (bylaws) [inline app/src/main/cpp/skyline/services/nvdrv/driver.h:33]: Fix this

**Review by bylaws — CHANGES_REQUESTED** [review]
Only a few nits

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvdevice.cpp (hunk 8 above)
Comment (PixelyIon) [reply app/src/main/cpp/skyline/services/nvdrv/devices/nvdevice.cpp:13]: I think it's essentially only done for services or nvdrv devices. I think we need a good bit more before we do this in common.

Comment (bylaws) [reply app/src/main/cpp/skyline/services/nvdrv/devices/nvdevice.cpp:13]: Oh I forgot std function doesn't actually have the name, ignore me

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp (hunk 9 above)
Comment (PixelyIon) [reply app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp:91]: Unused at the moment, might be used later on.

**Review by bylaws — APPROVED** [review]

**Pull request merged**